File: CrackingTheCodeInterview/3.2/MinStack.cs
/*	Program: Implement a stack which contains a min method which retreives min element of stack in constant time
 *	Author: Ankit Srivastava
 *	
 *	Date: 12/27/2015
 */

using System;

namespace CrackingTheCodingInterview
{
    /// <summary>
    /// Stack implementing an extra 'MinElement' method which retrieves the minimum element of stack in constant time
    /// </summary>
    public class MinStack
    {
        private SinglyLinkedList stack;

        private object minElement;

        /// <summary>
        /// Constructor which initializes the array implementing the stack
        /// </summary>
        public MinStack()
        {
            stack = new SinglyLinkedList();

            minElement = null;
        }

        /// <summary>
        /// <param name="d">Element 'd' to be pushed</param>
        /// Pushing an element on stack (twice in case element is l.t.e current min element) 
        /// </summary>
        public void Push(object d)
        {
            if (minElement == null || (int)d <= (int)minElement)
            {
                stack.Insert(minElement);
                stack.Insert(d);
                minElement = d;
            }
            else
            {
                stack.Insert(d);
            }
        }

        /// <summary>
        /// Pops an element off the stack. If its the min element, pop happens twice as we pushed twice. 
        /// This gives way to the next min element on the stack (which is pushed twice as well).
        /// <returns>Pops element from stack</returns>
        /// </summary>
        public object Pop()
        {
            object d = stack.DeleteHead();

            if (minElement.Equals(d))
            {
                minElement = stack.DeleteHead();
            }

            return d;
        }

        /// <summary>
        /// Peeks the top element of stack
        /// <returns>Top element of stack</returns>
        /// </summary>
        public object Peek()
        {
            return stack.PeekHead();
        }

        /// <summary>
        /// Returns the min element of stack
        /// <returns>Min element of stack</returns>
        /// </summary>
        public object Min()
        {
            return minElement;
        }

        /// <summary>
        /// Prints the linked list
        /// </summary>
        internal void Print()
        {
            stack.Print();
        }
    }

    internal class MinStackTest
    {
        static void Main()
        {
            MinStack minStack = new MinStack();

            minStack.Push(2);
            Console.WriteLine("Pushing 2 Min: {0}", minStack.Min());
            minStack.Push(6);
            Console.WriteLine("Pushing 6 Min: {0}", minStack.Min());
            minStack.Push(4);
            Console.WriteLine("Pushing 4 Min: {0}", minStack.Min());
            minStack.Push(1);
            Console.WriteLine("Pushing 1 Min: {0}", minStack.Min());
            minStack.Push(5);
            Console.WriteLine("Pushing 5 Min: {0}", minStack.Min());
            minStack.Push(1);
            Console.WriteLine("Pushing 1 Min: {0}", minStack.Min());

            minStack.Print();

            Console.WriteLine("Pop: {0} Min: {1}", minStack.Pop(), minStack.Min());
            minStack.Print();
            Console.WriteLine("Pop: {0} Min: {1}", minStack.Pop(), minStack.Min());
            minStack.Print();
            Console.WriteLine("Pop: {0} Min: {1}", minStack.Pop(), minStack.Min());
            minStack.Print();

            Console.WriteLine("Min: {0}", minStack.Min());

            Console.ReadLine();
        }
    }
}

File: CrackingTheCodeInterview/3.2/MinStack.cpp
#include <iostream>

using namespace std;

struct Node
{
	int data;
	Node *next;
};

class Stack
{
private:
	Node *head;
public:
	Stack();
	void Push(int d);
	int Pop();
	int Peek();
	bool IsEmpty();
};

Stack::Stack()
{
	head = NULL;
}

void Stack::Push(int d)
{
	Node *newNode = new Node();
	newNode->data = d;
	newNode->next = NULL;

	if (head != NULL)
	{
		newNode->next = head;
	}

	head = newNode;
}

int Stack::Pop()
{
	if (head == NULL)
	{
		cout << "Stack is empty";
		return -1;
	}
	else
	{
		int d = head->data;
		head = head->next;
		return d;
	}
}

int Stack::Peek()
{
	if (head == NULL)
	{
		cout << "Stack is empty";
		return -1;
	}
	else
	{
		return head->data;
	}
}

bool Stack::IsEmpty()
{
	return head == NULL;
}

class MinStack : public Stack
{
private:
	Stack min;
public:
	void Push(int d);
	int Pop();
	int GetMin();
};

void MinStack::Push(int d)
{
	if (min.IsEmpty() || (!min.IsEmpty() && min.Peek() > d))
	{
		min.Push(d);
	}

	Stack::Push(d);
}

int MinStack::Pop()
{
	if (Stack::IsEmpty())
	{
		cout << "Stack is empty";
		return -1;
	}
	else
	{
		int d = Stack::Pop();
		if (!min.IsEmpty() && d == min.Peek())
		{
			min.Pop();
		}

		return d;
	}
}

int MinStack::GetMin()
{
	return min.Peek();
}

int main()
{
	MinStack s;

	s.Push(1);
	s.Push(2);
	s.Push(0);
	s.Push(4);
	s.Push(5);

	cout << "Pop: " << s.Pop() << endl;
	cout << "Min: " << s.GetMin() << endl;
	cout << "Pop: " << s.Pop() << endl;
	cout << "Pop: " << s.Pop() << endl;
	
	cout << "Min: " << s.GetMin() << endl;

	return 0;
}